### ISSUE TITLE
Upload/Write Exception on MacOSX

### ISSUE THREAD

**Issue by LSOFT123** [hierynomus/smbj#78]
Hi,

Since your last fixes everything looks good under MacOSX except upload of file(s) with specific size. For instance, uploading a simple file sample2009q1a.ajpg (size = 4,221,390 bytes) returns the following exception at the end of transfer. Uploading the same file under Windows works, uploading another file under MacOSX with another size works.

```
[Packet Reader for macbookpro.home] c.h.s.t.PacketReader: Received packet << SMB2_CREATE with message id << 27 >> >>
[Packet Reader for macbookpro.home] c.h.s.c.Connection: Server granted us 2 credits for SMB2_CREATE with message id << 27 >>, now available: 538 credits
[Packet Reader for macbookpro.home] c.h.s.c.Connection: Send/Recv of packet with message id << 27 >> took << 2 ms >>
[Packet Reader for macbookpro.home] c.h.p.c.c.Promise: Setting << 27 >> to `SMB2_CREATE with message id << 27 >>`
c.h.s.s.File: Writing to sync1\sample2009q1a.ajpg from offset 0
c.h.s.c.Connection: Granted 64 (out of 538) credits to null with message id << 0 >>
c.h.s.t.t.DirectTcpTransport: Acquiring write lock to send packet << null with message id << 0 >> >>
c.h.s.t.t.DirectTcpTransport: Writing packet null with message id << 0 >>
FATAL EXCEPTION: Packet Reader for macbookpro.home
 java.lang.RuntimeException: com.hierynomus.smbj.transport.TransportException: java.io.EOFException: EOF while reading packet
	 at com.hierynomus.smbj.transport.PacketReader.run(PacketReader.java:53)
	 at java.lang.Thread.run(Thread.java:818)
  Caused by: com.hierynomus.smbj.transport.TransportException: java.io.EOFException: EOF while reading packet
	 at com.hierynomus.smbj.transport.tcp.DirectTcpPacketReader.readFully(DirectTcpPacketReader.java:80)
	 at com.hierynomus.smbj.transport.tcp.DirectTcpPacketReader._readTcpHeader(DirectTcpPacketReader.java:67)
	 at com.hierynomus.smbj.transport.tcp.DirectTcpPacketReader.doRead(DirectTcpPacketReader.java:56)
	 at com.hierynomus.smbj.transport.PacketReader.readPacket(PacketReader.java:68)
	 at com.hierynomus.smbj.transport.PacketReader.run(PacketReader.java:45)
	 at java.lang.Thread.run(Thread.java:818) 
  Caused by: java.io.EOFException: EOF while reading packet
	 at com.hierynomus.smbj.transport.tcp.DirectTcpPacketReader.readFully(DirectTcpPacketReader.java:80) 
	 at com.hierynomus.smbj.transport.tcp.DirectTcpPacketReader._readTcpHeader(DirectTcpPacketReader.java:67) 
	 at com.hierynomus.smbj.transport.tcp.DirectTcpPacketReader.doRead(DirectTcpPacketReader.java:56) 
	 at com.hierynomus.smbj.transport.PacketReader.readPacket(PacketReader.java:68) 
	 at com.hierynomus.smbj.transport.PacketReader.run(PacketReader.java:45) 
```

**Comment by pepijnve** [comment]
This particular error happens when the server unexpectedly closes the underlying TCP connection from its end. Could you provide a code snippet that shows what you're doing exactly? I can try to reproduce this on my own, but having some reproduction code to start from helps a lot in solving problems quickly.

**Comment by LSOFT123** [comment]
Code that uploads from FileInputStream:

```
public boolean storeFile(String path, InputStream in)
{
	boolean store = false;
	if ((path != null) && (diskShare != null))
	{
		com.hierynomus.smbj.share.File file = null;
		try 
		{
			String rPath = fixPath(path); // Clean remote path
			file = diskShare.openFile(rPath, EnumSet.of(AccessMask.GENERIC_WRITE), SMB2CreateDisposition.FILE_OVERWRITE_IF);
			file.write(new InputStreamByteChunkProvider(in));
			store = true;
		}
		catch (Exception e)
		{
			LOGGER.error(e.getMessage(),e);
		}
		finally
		{
			try
			{
				if (file != null) file.close();
			} 
			catch (Exception e)
			{
				LOGGER.error(e.getMessage(),e);
			}
		}
	}
	return store;
}
```

**Comment by pepijnve** [comment]
Tested this against my mac (macOS 10.12.5). For me the upload does work correctly, but I also get the same error as you do. For me this happens when the connection is closed. It seems that the macOS SMB server closes the TCP connection when the last session for that connection is logged off. To be more specific I see the following in Wireshark:
```
C -> S: Session logoff request
C <- S: Session logoff response
C -> S: TCP ACK
C <- S: TCP FIN/ACK
... remainder of TCP connection teardown follows
```

The smbj code currently doesn't expect this to happen and ends up throwing a RuntimeException because of it. I was looking into how this case could be dealt with easily but due to the asynchronous style in which the code is written it's a bit tricky without some ugly hacks.

@hierynomus the direction I was thinking was adding a 'closing connection' flag to PacketReader. When set it would ignore EOF errors. Would it make sense to add that?

**Comment by hierynomus** [comment]
@pepijnve What you're describing is something else than the initial bug report it seems, as @LSOFT123 claims that uploading a specific (size) file fails, but another succeeds fully.

**Comment by LSOFT123** [comment]
Here is the file I'm using if it can help. 
I'm not closing the session until the code I sent returns.

![sample2009qa](https://user-images.githubusercontent.com/29181341/27051909-2ab11f14-4fb8-11e7-9728-35957ce1a055.jpg)

**Comment by pepijnve** [comment]
@hierynomus I'm 99% sure this is the same issue. I get the exact same results as @LSOFT123 running the code he shared. Due to PacketReader running in a separate Thread and debug logging getting printed to stdout and fatal to stderr, the exact point at which the RuntimeException gets printed to the console varies. Sometimes it's after the last write debug logging, sometimes after the connection close logging. The packet capture I'm seeing in Wireshark is always identical.

@LSOFT123 could you try removing the session (or connection) close calls from your code to see if that makes the error go away? And could you check if the file is copied correctly or not (i.e., compare checksums)?
Removing the close call is not a fix for the issue of course, but just confirms (or not) if we're talking about the same thing.

**Comment by LSOFT123** [comment]
session.close() removed after transfer but it fails exactly the same.
The file copied is 0 bytes for both (under MacOSX).

**Comment by pepijnve** [comment]
So then it is something else after all... Are you familiar with either Wireshark or tcpdump by any chance? Getting a pcap file containing the back and forth between smbj and the server would be helpful again.

**Comment by LSOFT123** [comment]
Here are the SMB2 packets:
![smb2traces](https://user-images.githubusercontent.com/29181341/27251579-23dd0898-534a-11e7-8276-368254c8eaa8.png)

**Comment by pepijnve** [comment]
That second screenshot is mysterious. The sequence leading up to the gray line show the server acknowledging a bunch of packets and then terminating the TPC connection.
Could you export this out of Wireshark as a pcapng file and attach that? Then I can have a more in depth look.

**Comment by LSOFT123** [comment]
Here are 2 exports for the same upload (2 attempts):
[smb2traces-export.pcapng.gz](https://github.com/hierynomus/smbj/files/1082984/smb2traces-export.pcapng.gz)
[smb2traces2-export.pcapng.gz](https://github.com/hierynomus/smbj/files/1082985/smb2traces2-export.pcapng.gz)

Here is also the same file upload (that works) through a regular Windows 10 connected to MacOSX share:
[smb2traces-win10-export.pcapng.gz](https://github.com/hierynomus/smbj/files/1082991/smb2traces-win10-export.pcapng.gz)

**Comment by hierynomus** [comment]
@LSOFT123 Quick recap, what are the source and target OS's of each of the scenarios?

**Comment by LSOFT123** [comment]
Source is Java code running in Java Runtime under Windows 10. Target is MacOSX.

**Comment by LSOFT123** [comment]
I've just downloaded current source code from master and upload that failed before (under MaxOSX) works now. What did you fix?

**Comment by hierynomus** [comment]
I have no idea. Though I'm interested to know why it failed. Could you do a
"git bisect"  to see what was fixed and where?

Op 20 jun. 2017 8:03 p.m. schreef "LSOFT123" <notifications@github.com>:

**Comment by LSOFT123** [comment]
Not easy to say, lot of changes between today (June 20th) and June 12nd (date I downloaded code to report the issue).

**Comment by pepijnve** [comment]
With `git bisect` you should be able to pinpoint it fairly quickly. See https://git-scm.com/book/en/v2/Git-Tools-Debugging-with-Git#_binary_search for details.

**Comment by hierynomus** [comment]
@LSOFT123 Any updates on what caused the fix to appear all of a sudden?

**Comment by LSOFT123** [comment]
I'm not sure to have enough skills to do so.

**Comment by hierynomus** [comment]
Released 0.1.0 which contains these fixes, closing the issue
